File: agent/src/goal-runner/notebook.ts
/**
 * Notebook — lightweight cross-replan data persistence.
 *
 * Simple key-value store for data the agent discovers during execution
 * (prices, URLs, confirmation numbers, observations). Persists across
 * replans so that strategy changes don't lose discovered data.
 *
 * Injected into the LLM prompt as a single compact line when non-empty:
 *   SAVED DATA: cheapest=$149, dates=Mar15-17, booking_url=hotels.com/room/123
 *
 * Max 10 entries, capped at ~50 tokens in the prompt.
 */

// ─── Types ──────────────────────────────────────────────────────────────────

export type NotebookCategory = "data" | "url" | "observation" | "error";

export interface NotebookEntry {
  key: string;
  value: string;
  source: string;        // e.g. "step-5", "replan-2"
  category: NotebookCategory;
  timestamp: number;
}

// ─── Constants ──────────────────────────────────────────────────────────────

/** Maximum number of entries. Oldest entries evicted when exceeded. */
const MAX_ENTRIES = 10;

// ─── Notebook ───────────────────────────────────────────────────────────────

export class Notebook {
  private entries: Map<string, NotebookEntry> = new Map();

  /** Write or update an entry. Upserts by key. */
  write(key: string, value: string, source: string, category: NotebookCategory): void {
    this.entries.set(key, { key, value, source, category, timestamp: Date.now() });

    // Evict oldest if over max
    if (this.entries.size > MAX_ENTRIES) {
      let oldestKey: string | null = null;
      let oldestTime = Infinity;
      for (const [k, entry] of this.entries) {
        if (entry.timestamp < oldestTime) {
          oldestTime = entry.timestamp;
          oldestKey = k;
        }
      }
      if (oldestKey) this.entries.delete(oldestKey);
    }
  }

  /** Read a single entry by key. */
  read(key: string): string | undefined {
    return this.entries.get(key)?.value;
  }

  /** Get all entries as an array. */
  all(): NotebookEntry[] {
    return Array.from(this.entries.values());
  }

  /** Number of entries. */
  get size(): number {
    return this.entries.size;
  }

  /** Whether the notebook has any entries. */
  get isEmpty(): boolean {
    return this.entries.size === 0;
  }

  /**
   * Format for LLM prompt injection — single compact line.
   * Returns empty string if no entries.
   *
   * Example: "SAVED DATA: cheapest=$149, dates=Mar15-17, url=hotels.com/search"
   */
  toPromptContext(): string {
    if (this.entries.size === 0) return "";

    const pairs = Array.from(this.entries.values())
      .map(e => `${e.key}=${e.value}`)
      .join(", ");

    return `SAVED DATA: ${pairs}`;
  }

  /**
   * Format for GoalResult summary — grouped by category.
   */
  toSummary(): string {
    if (this.entries.size === 0) return "";

    const byCategory = new Map<NotebookCategory, NotebookEntry[]>();
    for (const entry of this.entries.values()) {
      const list = byCategory.get(entry.category) ?? [];
      list.push(entry);
      byCategory.set(entry.category, list);
    }

    const parts: string[] = [];
    for (const [category, items] of byCategory) {
      const label = category.charAt(0).toUpperCase() + category.slice(1);
      const formatted = items.map(e => `  ${e.key}: ${e.value}`).join("\n");
      parts.push(`${label}:\n${formatted}`);
    }

    return parts.join("\n");
  }

  /** Deep-copy entries for checkpoint storage. */
  snapshot(): NotebookEntry[] {
    return Array.from(this.entries.values()).map(e => ({ ...e }));
  }

  /** Restore from a checkpoint snapshot. */
  restoreFromSnapshot(entries: NotebookEntry[]): void {
    this.entries.clear();
    for (const entry of entries) {
      this.entries.set(entry.key, { ...entry });
    }
  }

  /** Clear all entries. */
  clear(): void {
    this.entries.clear();
  }
}

File: cel/cel-goal-runner/src/notebook.rs
//! Notebook — key-value store that survives replans.
//!
//! Records data discovered during execution (prices, URLs, confirmation numbers).
//! Persists across T2/T3 replanning tiers. Restored from checkpoints on T3 backtrack.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Persistent notebook for cross-replan data.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Notebook {
    entries: HashMap<String, NotebookEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotebookEntry {
    pub value: String,
    pub category: String,
    pub step_recorded: u32,
}

impl Notebook {
    pub fn new() -> Self {
        Self::default()
    }

    /// Write a key-value pair to the notebook.
    pub fn write(&mut self, key: &str, value: &str, category: &str, step: u32) {
        self.entries.insert(key.to_string(), NotebookEntry {
            value: value.to_string(),
            category: category.to_string(),
            step_recorded: step,
        });
    }

    /// Read a value from the notebook.
    pub fn read(&self, key: &str) -> Option<&str> {
        self.entries.get(key).map(|e| e.value.as_str())
    }

    /// Get all entries as a formatted string for injection into prompts.
    pub fn to_prompt_context(&self) -> String {
        if self.entries.is_empty() {
            return String::new();
        }
        let mut lines = vec!["## Notebook (data discovered so far)".to_string()];
        for (key, entry) in &self.entries {
            lines.push(format!("- {key}: {} [{}]", entry.value, entry.category));
        }
        lines.join("\n")
    }

    /// Snapshot for checkpoint restore.
    pub fn snapshot(&self) -> HashMap<String, NotebookEntry> {
        self.entries.clone()
    }

    /// Restore from a checkpoint snapshot.
    pub fn restore(&mut self, snapshot: HashMap<String, NotebookEntry>) {
        self.entries = snapshot;
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    pub fn len(&self) -> usize {
        self.entries.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_write_read() {
        let mut nb = Notebook::new();
        nb.write("price", "$149", "data", 3);
        assert_eq!(nb.read("price"), Some("$149"));
        assert_eq!(nb.len(), 1);
    }

    #[test]
    fn test_overwrite() {
        let mut nb = Notebook::new();
        nb.write("price", "$149", "data", 3);
        nb.write("price", "$129", "data", 5);
        assert_eq!(nb.read("price"), Some("$129"));
    }

    #[test]
    fn test_snapshot_restore() {
        let mut nb = Notebook::new();
        nb.write("url", "https://example.com", "nav", 1);
        nb.write("price", "$99", "data", 2);

        let snap = nb.snapshot();
        nb.write("extra", "gone", "data", 3);
        assert_eq!(nb.len(), 3);

        nb.restore(snap);
        assert_eq!(nb.len(), 2);
        assert!(nb.read("extra").is_none());
        assert_eq!(nb.read("price"), Some("$99"));
    }

    #[test]
    fn test_prompt_context() {
        let mut nb = Notebook::new();
        nb.write("hotel", "Marriott Amsterdam", "data", 1);
        let ctx = nb.to_prompt_context();
        assert!(ctx.contains("Notebook"));
        assert!(ctx.contains("Marriott Amsterdam"));
    }
}
